File: src/pieces.js
import { Position } from "chessboard/model/Position";

export function getPawnMoves(startCoords, team, board) {
  return pathMarchPotentialMoves(startCoords, team, 1, true, true, board);
}

export function getBishopMoves(startCoords, team, board) {
  return pathMarchPotentialMoves(startCoords, team, 7, true, false, board);
}

export function getRookMoves(startCoords, team, board) {
  return pathMarchPotentialMoves(startCoords, team, 7, false, true, board);
}

export function getQueenMoves(startCoords, team, board) {
  return pathMarchPotentialMoves(startCoords, team, 7, true, true, board);
}

export function getKingMoves(startCoords, team, board) {
  // TODO: add support for checkmate
  return getPawnMoves(startCoords, team, board);
}

export function getKnightMoves(
  startCoords,
  team,
  board,
) {
  const [x, y] = startCoords;
  const potentialMoves = [
    [x + 1, y + 2],
    [x - 1, y + 2],
    [x + 1, y - 2],
    [x - 1, y - 2],
    [x + 2, y + 1],
    [x - 2, y + 1],
    [x + 2, y - 1],
    [x - 2, y - 1],
  ]

  return potentialMoves.filter((m) => {
    if (!moveInBounds(board, m)) {
      return false
    }
    const possibleSquare = Position.coordinatesToSquare(m);
    const possiblePiece = board.getPiece(possibleSquare);
    if (possiblePiece) {
      return getTeam(possiblePiece) !== team;
    } else {
      return true;
    }
  });

}

// TODO(sjayakar): I would like a test
export function pathMarchPotentialMoves(
  startCoords,
  team,
  // 1 -> 7; not a requirement
  maxDistance,
  allowDiagonals,
  allowStraight,
  // TODO(sjayakar): consider passing in all the pieces. something to
  // think about is some pieces _might_ not block movement.
  //
  // So what I need is
  // - board width + height for boundary
  // - location of all the pieces
  board
) {
  const [startX, startY] = startCoords;
  const possibleCoords = [];

  if (allowDiagonals) {
    // If any of the directions get blocked, don't add any more moves
    // in that direction
    let TLBlocked = false;
    let TRBlocked = false;
    let BLBlocked = false;
    let BRBlocked = false;
    for (let i = 1; i <= maxDistance; ++i) {
      TLBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX - i, startY + i],
        TLBlocked
      );
      TRBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX + i, startY + i],
        TRBlocked
      );
      BLBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX - i, startY - i],
        BLBlocked
      );
      BRBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX + i, startY - i],
        BRBlocked
      );
    }
  }
  if (allowStraight) {
    // If any of the directions get blocked, don't add any more moves
    // in that direction
    let upBlocked = false;
    let downBlocked = false;
    let leftBlocked = false;
    let rightBlocked = false;
    for (let i = 1; i <= maxDistance; ++i) {
      upBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX, startY + i],
        upBlocked
      );
      downBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX, startY - i],
        downBlocked
      );
      leftBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX - i, startY],
        leftBlocked
      );
      rightBlocked = potentiallyAddDirection(
        board,
        team,
        possibleCoords,
        [startX + i, startY],
        rightBlocked
      );
    }
  }
  return possibleCoords;
}

function moveInBounds(
  board,
  move,
) {
  const [x, y] = move;
  return !(x < 0 || x >= board.props.boardWidth || y < 0 || y >= board.props.boardHeight);
}

// mutates move list, returns blocked
function potentiallyAddDirection(
  board,
  team,
  possibleMoves,
  possibleMove,
  blocked
) {
  if (blocked) {
    return blocked;
  }

  if (!moveInBounds(board, possibleMove)) {
    return blocked;
  }

  const possibleSquare = Position.coordinatesToSquare(possibleMove);

  const possiblePiece = board.getPiece(possibleSquare);
  if (possiblePiece) {
    // Regardless, both of these will block the piece from
    // future moves. However, if it's an enemy piece, we can
    // still move there.
    blocked = true;
    if (getTeam(possiblePiece) !== team) {
      possibleMoves.push(possibleMove);
    }
  } else {
    possibleMoves.push(possibleMove);
  }

  return blocked;
}

export function getTeam(piece) {
  return piece[0];
}

File: python/pieces.py
from typing import List, Tuple, Optional
from enum import Enum

Position = Tuple[int, int]

class Team(Enum):
    WHITE = 'w'
    BLACK = 'b'

    def opposite(self):
        return Team.BLACK if self == Team.WHITE else Team.WHITE

class PieceType(Enum):
    PAWN = 'p'
    KNIGHT = 'n'
    BISHOP = 'b'
    ROOK = 'r'
    QUEEN = 'q'
    KING = 'k'

def get_pawn_moves(board, position: Position, team: Team, board_width: int, board_height: int) -> List[Position]:
    moves = []
    row, col = position
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
    for dr, dc in directions:
        new_pos = (row + dr, col + dc)
        if is_in_bounds(new_pos, board_width, board_height):
            target_piece = get_piece_at(board, new_pos)
            if target_piece is None or target_piece.team != team:
                moves.append(new_pos)
    return moves

def get_knight_moves(board, position: Position, team: Team, board_width: int, board_height: int) -> List[Position]:
    moves = []
    row, col = position
    deltas = [
        (-2, -1), (-2, 1),
        (-1, -2), (-1, 2),
        (1, -2),  (1, 2),
        (2, -1),  (2, 1)
    ]
    for dr, dc in deltas:
        new_pos = (row + dr, col + dc)
        if is_in_bounds(new_pos, board_width, board_height):
            target_piece = get_piece_at(board, new_pos)
            if target_piece is None or target_piece.team != team:
                moves.append(new_pos)
    return moves

def get_straight_line_moves(board, position: Position, team: Team, directions: List[Tuple[int, int]], 
                            max_distance: int, board_width: int, board_height: int) -> List[Position]:
    moves = []
    for dr, dc in directions:
        new_row, new_col = position
        for _ in range(max_distance):
            new_row += dr
            new_col += dc
            new_pos = (new_row, new_col)
            if not is_in_bounds(new_pos, board_width, board_height):
                break
            target_piece = get_piece_at(board, new_pos)
            if target_piece is None:
                moves.append(new_pos)
            else:
                if target_piece.team != team:
                    moves.append(new_pos)
                break
    return moves

def get_bishop_moves(board, position: Position, team: Team, board_width: int, board_height: int) -> List[Position]:
    directions = [(-1, -1), (-1, 1), (1, -1), (1, 1)]
    return get_straight_line_moves(board, position, team, directions, max_distance=7, 
                                   board_width=board_width, board_height=board_height)

def get_rook_moves(board, position: Position, team: Team, board_width: int, board_height: int) -> List[Position]:
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    return get_straight_line_moves(board, position, team, directions, max_distance=7, 
                                   board_width=board_width, board_height=board_height)

def get_queen_moves(board, position: Position, team: Team, board_width: int, board_height: int) -> List[Position]:
    directions = [
        (-1, -1), (-1, 1), (1, -1), (1, 1),
        (-1, 0), (1, 0), (0, -1), (0, 1)
    ]
    return get_straight_line_moves(board, position, team, directions, max_distance=7, 
                                   board_width=board_width, board_height=board_height)

def get_king_moves(board, position: Position, team: Team, board_width: int, board_height: int) -> List[Position]:
    moves = []
    row, col = position
    
    pawn_moves = get_pawn_moves(board, position, team, board_width, board_height)
    other_king_type = PieceType.KING
    other_team = team.opposite()
    
    for move in pawn_moves:
        neighboring_moves = get_pawn_moves(board, move, team, board_width, board_height)
        contains_king = False
        for nm in neighboring_moves:
            piece = get_piece_at(board, nm)
            if piece and piece.piece_type == other_king_type and piece.team == other_team:
                contains_king = True
                break
        if not contains_king:
            moves.append(move)
    
    return moves

def is_in_bounds(position: Position, board_width: int, board_height: int) -> bool:
    row, col = position
    return 0 <= row < board_height and 0 <= col < board_width

def get_piece_at(board, position: Position):
    row, col = position
    return board[row][col]
